### PR TITLE
Add logging instrumentation for bunyan

### PR DESCRIPTION
**Pull request by brianphillips**
## Proposed Release Notes

* Added instrumentation to bunyan to support application logging use cases: forwarding, local decorating, and metrics.
   
   Big thanks to @brianphillips for his contribution 🚀

## Details

The instrumentation added wraps two methods within bunyan. First, the `createLogger` method is wrapped to add usage metrics as well as log forwarding (a stream is added to the bunyan logger that passes raw JSON log lines to the APM agent's log aggregator). Second, the internal (but unchanged for 10+ years) `_emit` method is used for decorating log lines with necessary fields for associating each log line to an application and transaction within New Relic.

### CODE REVIEW

File: lib/instrumentation/bunyan.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+        const bunyanLogger = createLogger.apply(this, arguments)
+
+        if (logForwardingEnabled) {
+          bunyanLogger.addStream({
+            type: 'raw',
+            level: bunyanLogger.level(),
+            stream: {
+              write: function nrLogWrite(logLine) {
+                agent.logs.add(logLine)
```
Comment (bizob2828) [inline lib/instrumentation/bunyan.js:53]: so from my comments below I think this is where you would do all the log line manipulation for the forwarding use case

**Review by bizob2828 — COMMENTED** [review]
Hi @brianphillips thanks for the PR! You've helped us to a lot of the leg work around finding where to instrument.  

I'd link our spec but it's private but NR-LINKING metadata is a separate use case that is mutually exclusive from log forwarding and appending the linked metadata as keys on json. Our [docs](https://docs.newrelic.com/docs/logs/logs-context/configure-logs-context-nodejs/) explain the 3 use cases.  The first 2: log forwarding and local log decorating need to be built into this code along with the relevant logging metrics.  You did find our helpers which should set you up for success.  

It's also worth nothing we need versioned tests for these. I would look at [winston](https://github.com/newrelic/node-newrelic/tree/main/test/versioned/winston) and [pino](https://github.com/newrelic/node-newrelic/tree/main/test/versioned/pino) for how to do this.  

Please let us know if you have any questions but this is definitely a good first step.

File: lib/instrumentation/bunyan.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      }
+    })
+  }
+
+  const localDecoratingEnabled = isLocalDecoratingEnabled(config)
+
+  shim.wrap(bunyan.prototype, '_emit', function wrapEmit(shim, emit) {
+    return function wrappedEmit() {
+      const args = shim.argsToArray.apply(shim, arguments)
+      const rec = args[0] || {}
```
[the remark this answers is not in the document]
Comment (bizob2828) [reply lib/instrumentation/bunyan.js:68]: our jira integration tried to link that nr linking above, that was not intended to be a link anywhere

File: lib/instrumentation/bunyan.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      if (rec.timestamp) {
+        logger.traceOnce(
+          'Overwriting `timestamp` key; assigning original value to `original_timestamp`.'
+        )
+        rec.original_timestamp = rec.timestamp
+      }
+      rec.timestamp = Date.now()
+
+      // put log message into a consistent spot and ensure it's not too long
+      rec.message = truncate(rec.msg)
```
Comment (bizob2828) [inline lib/instrumentation/bunyan.js:88]:
```suggestion
if (isLocalDecoratingEnabled) {
      rec.message = rec.msg + agent.getNRLinkingMetadata()
   }
```

File: lib/instrumentation/bunyan.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+      rec.msg = '' // this allows the CLI pretty-printing to continue working, for the most part
```
Comment (bizob2828) [inline lib/instrumentation/bunyan.js:89]: you can keep this as is if you make all the log line manipulation in the add stream method above

File: lib/instrumentation/bunyan.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+          })
+        }
+        return bunyanLogger
+      }
+    })
+  }
+
+  const localDecoratingEnabled = isLocalDecoratingEnabled(config)
+
+  shim.wrap(bunyan.prototype, '_emit', function wrapEmit(shim, emit) {
```
Comment (bizob2828) [inline lib/instrumentation/bunyan.js:65]: I think this instrumentation should only count log lines when metrics are enabled and append the piped nr linking metadata when local log decorating is enabled.  The rest should go in the stream write handler above

**Comment by brianphillips** [comment]
@bizob2828 - Thanks so much for the quick feedback!

>  NR-LINKING metadata is a separate use case that is mutually exclusive from log forwarding and appending the linked metadata as keys on json

Just to confirm, appending the `NR-LINKING` metadata is the same as local decoration and is mutually exclusive from the log forwarding approach?

> It's also worth nothing we need versioned tests for these. I would look at [winston](https://github.com/newrelic/node-newrelic/tree/main/test/versioned/winston) and [pino](https://github.com/newrelic/node-newrelic/tree/main/test/versioned/pino) for how to do this.

Yes, wanted to get confirmation on the direction before I added tests.

Thanks again!

**Comment by bizob2828** [comment]
@brianphillips NR-LINKING metadata is the same as local log decorating. It is mutually exclusive from log forwarding but it doesn't have to be written s mutually exclusive. Just tested in versioned tests.

**Comment by bizob2828** [comment]
> @bizob2828 - Thanks so much for the quick feedback!
> 
> 
> 
> >  NR-LINKING metadata is a separate use case that is mutually exclusive from log forwarding and appending the linked metadata as keys on json
> 
> 
> 
> Just to confirm, appending the `NR-LINKING` metadata is the same as local decoration and is mutually exclusive from the log forwarding approach?

@brianphillips NR-LINKING metadata is local log decorating. It is mutually exclusive but doesn't have to be coded as such. It should just be tested separately. 
> 
> 
> 
> > It's also worth nothing we need versioned tests for these. I would look at [winston](https://github.com/newrelic/node-newrelic/tree/main/test/versioned/winston) and [pino](https://github.com/newrelic/node-newrelic/tree/main/test/versioned/pino) for how to do this.
> 
> 
> 
> Yes, wanted to get confirmation on the direction before I added tests.
> 
> 
> 
> Thanks again!

File: lib/instrumentation/bunyan.js (hunk 3 above)
Comment (brianphillips) [reply lib/instrumentation/bunyan.js:88]: Your suggested change would result in the `NR-LINKING` piped metadata being embedded inside the JSON object that is logged. Is that OK? Or does it need to be at the end just before the new-line character? Bunyan makes that quite difficult because it handles all the JSON stringification internally with non-exported (aka non-monkey-patchable) functions. The only viable option (that I can think of) would be to wrap the `write` method of each of the streams which is the point where we have a string that we can append to.

Bottom line question: is it an option to modify the fields of the log message when local decorating is enabled or should I go down the route of wrapping all the `stream.write` methods to inject the `NR-LINKING` piped metadata at that layer?

Comment (bizob2828) [reply lib/instrumentation/bunyan.js:88]: > Your suggested change would result in the NR-LINKING piped metadata being embedded inside the JSON object that is logged. Is that OK?

Yes that is the spec for this feature of local log decorating.  The log forwarding feature operates on a copy of the log line which is why I suggest putting the rest of this logic in the stream.write function.  So you can see [here](https://github.com/newrelic/node-newrelic/blob/main/lib/instrumentation/winston.js#L86) we edit the existing log line for local log decorating and use it to increment metrics.  Then [here](https://github.com/newrelic/node-newrelic/blob/main/lib/instrumentation/winston.js#L46) we do the rest of the  logic in a winston formatter which does not affect the original log line

**Comment by brianphillips** [comment]
@bizob2828 I have (locally passing) versioned tests with reasonably good coverage, I think. If there are edge cases you'd like to see covered, please point me in the right direction!

File: test/versioned/bunyan/bunyan.tap.js [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+const tap = require('tap')
+const helper = require('../../lib/agent_helper')
+const { validateLogLine } = require('../../lib/logging-helper')
+const { LOGGING } = require('../../../lib/metrics/names')
+// winston puts the log line getting construct through formatters on a symbol
```
Comment (bizob2828) [inline test/versioned/bunyan/bunyan.tap.js:12]: you have a bit of copy pasta in here. i've seen a few winston strings 😄

File: test/versioned/bunyan/helpers.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    write: function write(logLine) {
+      this.logs.push(logLine)
+    }
+  }
+}
+
+/**
+ * Log lines in and out of a transaction for every logger.
+ * @param {object} opts
+ * @param {DerivedLogger} opts.logger instance of winston
```
Comment (bizob2828) [inline test/versioned/bunyan/helpers.js:26]: this type and winston is from the winston helpers

File: test/versioned/bunyan/bunyan.tap.js [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+      t.notOk(msg.stack, 'stack should be removed')
+      t.notOk(msg.trace, 'trace should be removed')
+      t.end()
+    })
+  })
+
+  t.test('metrics', (t) => {
+    t.autoend()
+
+    t.test('should log unknown for fatal log level', (t) => {
```
Comment (bizob2828) [inline test/versioned/bunyan/bunyan.tap.js:197]: this should log accordingly, I think we missed adding this as a constant as I see pino support this too. You can just add it to the levels constants [here](https://github.com/newrelic/node-newrelic/blob/main/lib/metrics/names.js#L276) then you can just remove this test and include in the one below.

**Review by bizob2828 — CHANGES_REQUESTED** [review]
Great job figuring out the versioned tests. After reviewing the tests I realized we were missing some level constants. But once updated I will approve

File: test/versioned/bunyan/bunyan.tap.js (hunk 6 above)
Comment (brianphillips) [reply test/versioned/bunyan/bunyan.tap.js:12]: you caught me! :laughing: 

(I'll clean it up)

**Review by bizob2828 — DISMISSED** [review]
Thanks for the contribution @brianphillips!

**Comment by bizob2828** [comment]
The CI build is failing for unrelated reasons, I'll dig more in the morning

**Comment by bizob2828** [comment]
> The CI build is failing for unrelated reasons, I'll dig more in the morning

Oh, lol it is related but in a very roundabout way.  Since you've added support for bunyan and this test checks for metrics the "has this module loaded?" metrics for bunyan are now included. I'll push a fix

**Review by bizob2828 — APPROVED** [review]

**Comment by brianphillips** [comment]
@bizob2828 Thanks for all the assistance getting this merged! What's your release schedule like?

**Comment by bizob2828** [comment]
> @bizob2828 Thanks for all the assistance getting this merged! What's your release schedule like?

@brianphillips  We tend to avoid releasing on Friday so this will go out on Monday.
